File: frontend/src/lib/api/alertRules.ts
/**
 * Typed client for `/api/internal/alert-rules*` and `/alert-templates*`
 * (docs/API.md §5, roadmap slices 038/041). Shapes mirror the payloads
 * `backend/src/flightsite/api/alert_rules.py` builds — see its
 * `_rule_payload` / `_template_payload`.
 *
 * The condition document is the part worth reading carefully. The backend
 * validates it with the very model the *stored* document is parsed with
 * (`flightsite.alerts.model.RuleConditions`), so a document this client sends
 * and one it reads back are the same shape, and a rule the API accepts is by
 * construction one the engine can evaluate. That is what makes slice 041's
 * round-trip property ("rules created in the UI evaluate identically to
 * API-created rules") hold without this file re-deriving any of it.
 *
 * Optional members are written `?: T | null` rather than `?: T` because the
 * backend dumps with `exclude_none=True`: an unset condition is an *absent*
 * key on the way out, and either an absent key or an explicit `null` is
 * accepted on the way in.
 */
import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseMutationResult,
  type UseQueryResult,
} from "@tanstack/react-query";

import { apiFetch } from "@/lib/api/client";
import type { AlertSeverity } from "@/lib/api/sightings";

/** The mission categories a classification condition may require, spelled
 * exactly as `flightsite.classification.vocabulary.MissionCategory` does.
 * `unknown` is deliberately absent: the backend refuses it, because a rule
 * matching every airframe no metadata source has heard of would be a rule
 * about FlightSite's ignorance rather than about aircraft. */
export type AlertMissionCategory =
  | "commercial_passenger"
  | "cargo"
  | "general_aviation"
  | "business_aviation"
  | "military"
  | "government"
  | "law_enforcement"
  | "medical"
  | "firefighting"
  | "training"
  | "helicopter";

/** SPEC §39 classification requirements. The three flags are requirements,
 * never negations — `military: true` means "must be military" and
 * `military: false` means "do not care". There is no way to say "must not
 * be", because that is a boolean `NOT` and SPEC §43 limits v1 to `AND` over
 * positive conditions. */
export interface AlertClassificationCondition {
  military: boolean;
  government: boolean;
  law_enforcement: boolean;
  mission?: AlertMissionCategory | null;
}

/** A receiver-relative rarity threshold (SPEC §44). `max_sightings` is
 * inclusive, so `1` is exactly "never seen here before". */
export interface AlertRarityCondition {
  max_sightings: number;
}

/**
 * One rule's `AND`-combined condition set (docs/DATA_MODEL.md §4.2).
 *
 * Flat, with every member optional, because SPEC §43 gives v1 no nested
 * boolean trees: every condition present must hold. `version` is the
 * document's forward door and this build reads and writes `1` only.
 */
export interface AlertRuleConditions {
  version: 1;
  classification?: AlertClassificationCondition | null;
  /** Exact match on the resolved ICAO type designator, case-insensitively. */
  type_code?: string | null;
  /** Case-insensitive *substring* of the resolved model name — the stored
   * value is registry prose ("Boeing C-17A Globemaster III"), so a user
   * means "Globemaster", not that string character for character. */
  model?: string | null;
  /** Membership of one specific watchlist, by id. */
  watchlist_id?: number | null;
  /** Membership of *any* watchlist. Exists beside `watchlist_id` because a
   * template instantiated at first run has no watchlist to name yet. */
  watchlist_any?: boolean;
  rare_aircraft?: AlertRarityCondition | null;
  rare_type?: AlertRarityCondition | null;
  max_distance_nm?: number | null;
  min_distance_nm?: number | null;
  max_alt_ft?: number | null;
  min_alt_ft?: number | null;
  /** Whether the rule also applies to aircraft the decoder reports on the
   * ground. `false` — the default — is SPEC §40's "excluded from relevant
   * alerts". Not itself a condition: a rule that says only this matches
   * nothing, and the backend refuses it. */
  applies_on_ground?: boolean;
}

export interface AlertRule {
  id: number;
  name: string;
  description: string | null;
  severity: AlertSeverity;
  enabled: boolean;
  /** `null` for a user-written rule; the shipped template's key for one that
   * came from the gallery or from first-run instantiation. Provenance is not
   * replaceable — tuning a shipped rule does not make it stop having been
   * shipped. */
  template_key: string | null;
  conditions: AlertRuleConditions;
  /** The rule stated in prose, one phrase per condition, composed by the
   * backend's `RuleConditions.describe()`. Rendered as-is so every client
   * says the same thing about a rule rather than re-implementing it. */
  describes: string[];
  created_at: string;
  updated_at: string;
}

/** One shipped template (SPEC §45). `builtin` is the field that matters: a
 * built-in template describes behaviour that is already on and cannot be
 * turned off (SPEC §47's emergency squawks), so its `conditions` are `null`
 * because there is no rule to create. */
export interface AlertTemplate {
  key: string;
  name: string;
  description: string;
  severity: AlertSeverity;
  builtin: boolean;
  conditions: AlertRuleConditions | null;
}

export interface AlertRulesListResponse {
  rules: AlertRule[];
}

export interface AlertTemplatesListResponse {
  templates: AlertTemplate[];
}

/** `POST` and `PUT` body. One shape for both, because `PUT` is a full
 * replace rather than a patch — a partial update of an `AND` condition set
 * is ambiguous about whether an omitted condition was meant to be removed. */
export interface AlertRuleWriteInput {
  name: string;
  description: string | null;
  severity: AlertSeverity;
  conditions: AlertRuleConditions;
  enabled: boolean;
}

const RULES_PATH = "/api/internal/alert-rules";
const TEMPLATES_PATH = "/api/internal/alert-templates";

const JSON_HEADERS = { "Content-Type": "application/json" };

export function getAlertRules(): Promise<AlertRulesListResponse> {
  return apiFetch<AlertRulesListResponse>(RULES_PATH);
}

export function getAlertTemplates(): Promise<AlertTemplatesListResponse> {
  return apiFetch<AlertTemplatesListResponse>(TEMPLATES_PATH);
}

export function createAlertRule(input: AlertRuleWriteInput): Promise<AlertRule> {
  return apiFetch<AlertRule>(RULES_PATH, {
    method: "POST",
    headers: JSON_HEADERS,
    body: JSON.stringify(input),
  });
}

export function updateAlertRule(
  ruleId: number,
  input: AlertRuleWriteInput,
): Promise<AlertRule> {
  return apiFetch<AlertRule>(`${RULES_PATH}/${ruleId}`, {
    method: "PUT",
    headers: JSON_HEADERS,
    body: JSON.stringify(input),
  });
}

export function deleteAlertRule(ruleId: number): Promise<void> {
  return apiFetch<void>(`${RULES_PATH}/${ruleId}`, { method: "DELETE" });
}

/** Turns one shipped template into a rule carrying its `template_key`. The
 * body is empty on purpose: the conditions and severity come from the
 * backend's catalogue, never from here, so provenance stays a statement the
 * server makes rather than a claim a client attaches. */
export function instantiateAlertTemplate(key: string): Promise<AlertRule> {
  return apiFetch<AlertRule>(`${TEMPLATES_PATH}/${key}/rules`, {
    method: "POST",
  });
}

/** Query key for the rule list. */
export const alertRulesQueryKey = ["alert-rules"] as const;

/** Query key for the shipped-template catalogue. */
export const alertTemplatesQueryKey = ["alert-templates"] as const;

export function useAlertRulesQuery(): UseQueryResult<AlertRulesListResponse> {
  return useQuery({ queryKey: alertRulesQueryKey, queryFn: getAlertRules });
}

/** The shipped catalogue is static data compiled into the backend — it
 * cannot change while the app is running — so it is never refetched for
 * staleness. What *does* change is which templates have been instantiated,
 * and that is read from the rule list's `template_key`, not from here. */
export function useAlertTemplatesQuery(): UseQueryResult<AlertTemplatesListResponse> {
  return useQuery({
    queryKey: alertTemplatesQueryKey,
    queryFn: getAlertTemplates,
    staleTime: Infinity,
  });
}

/** Every rule mutation invalidates the rule list, so no consumer has to
 * remember to refetch: the list is what the rules tab renders *and* what the
 * template gallery reads to decide which templates are already instantiated. */
function invalidateAlertRules(
  queryClient: ReturnType<typeof useQueryClient>,
): void {
  void queryClient.invalidateQueries({ queryKey: alertRulesQueryKey });
}

export function useCreateAlertRuleMutation(): UseMutationResult<
  AlertRule,
  Error,
  AlertRuleWriteInput
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: createAlertRule,
    onSuccess: () => {
      invalidateAlertRules(queryClient);
    },
  });
}

export function useUpdateAlertRuleMutation(): UseMutationResult<
  AlertRule,
  Error,
  { ruleId: number; input: AlertRuleWriteInput }
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ ruleId, input }) => updateAlertRule(ruleId, input),
    onSuccess: () => {
      invalidateAlertRules(queryClient);
    },
  });
}

export function useDeleteAlertRuleMutation(): UseMutationResult<
  void,
  Error,
  number
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: deleteAlertRule,
    onSuccess: () => {
      invalidateAlertRules(queryClient);
    },
  });
}

export function useInstantiateAlertTemplateMutation(): UseMutationResult<
  AlertRule,
  Error,
  string
> {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: instantiateAlertTemplate,
    onSuccess: () => {
      invalidateAlertRules(queryClient);
    },
  });
}

File: backend/src/flightsite/api/alert_rules.py
"""Internal alert-rule CRUD: ``/api/internal/alert-rules`` and ``/alert-templates``.

A router module of its own, mounted from :mod:`flightsite.api.internal` with a
single ``include_router`` line. That is deliberate and not merely tidy: the
internal surface is a shared file that several slices extend at once, and a
slice that appends a few hundred lines to the end of it collides with every
other slice doing the same. One line in the shared file and everything else
here means this slice's additions and the next one's cannot touch the same
region. The mount keeps ``include_in_schema=False`` from the app-level
inclusion (ADR-0007, ``docs/API.md`` §5), so nothing here reaches the published
OpenAPI document.

Where validation happens, and where it does not
------------------------------------------------

Shape validation is Pydantic's, in :mod:`flightsite.alerts.schemas`, and the
condition set is validated by the very model the stored document is parsed with
(:class:`flightsite.alerts.model.RuleConditions`). So a rule this endpoint
accepts is by construction one the engine can evaluate — including the
threshold bounds and the two "this rule can never match anything" checks (an
empty condition set, an inverted distance or altitude window), which are
validation errors at the moment a rule is written rather than a mystery at the
moment it fails to fire.

*Domain* rules — a blank name, a rule id that does not exist — belong to
:class:`flightsite.alerts.service.AlertService`, which applies them identically
whichever caller reaches it, and they arrive here as
:mod:`flightsite.alerts.errors` exceptions to be mapped onto statuses. The same
split :mod:`flightsite.watchlists` draws.

Every mutation goes through the service, which recompiles the engine's rule set
before returning — so a client that just created a rule sees the live picture
reflect it on its very next read, with no propagation delay to reason about.
That is the same guarantee the watchlist endpoints give, and slice 041's
round-trip test ("rules created in the UI evaluate identically to API-created
rules") rests on it.
"""

from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Body, HTTPException, Request, status

from flightsite.alerts import (
    SHIPPED_TEMPLATES,
    AlertRuleNotFoundError,
    AlertRuleRecord,
    AlertRuleValueError,
    AlertRuleWriteRequest,
    AlertService,
    AlertTemplate,
)
from flightsite.api.serializers import iso_utc
from flightsite.db import from_epoch_ms

router = APIRouter()


def _service(request: Request) -> AlertService:
    service: AlertService = request.app.state.alerts
    return service


def _rule_payload(record: AlertRuleRecord) -> dict[str, Any]:
    """One rule as the Alerts page's list/detail row (``docs/API.md`` §5).

    ``conditions`` is echoed as the validated document rather than as the
    stored text, so a client always reads the same shape it would have to send
    back — and ``describes`` beside it is the rule stated in prose, which is
    what a list row shows under the name without every client re-implementing
    :meth:`~flightsite.alerts.model.RuleConditions.describe`.
    """
    return {
        "id": record.id,
        "name": record.name,
        "description": record.description,
        "severity": record.severity.value,
        "enabled": record.enabled,
        "template_key": record.template_key,
        "conditions": record.conditions.model_dump(mode="json", exclude_none=True),
        "describes": list(record.conditions.describe()),
        "created_at": iso_utc(from_epoch_ms(record.created_ms)),
        "updated_at": iso_utc(from_epoch_ms(record.updated_ms)),
    }


def _template_payload(template: AlertTemplate) -> dict[str, Any]:
    """One shipped template, as the template gallery reads it (SPEC §45).

    ``builtin`` is the field that matters: a built-in template describes
    behaviour that is already on and cannot be turned off (SPEC §47's emergency
    squawks), so a gallery must render it as a statement rather than as a
    switch. Its ``conditions`` are ``null`` because there is no rule to create.
    """
    conditions = template.conditions
    return {
        "key": template.key,
        "name": template.name,
        "description": template.description,
        "severity": template.severity.value,
        "builtin": template.builtin,
        "conditions": (
            None if conditions is None else conditions.model_dump(mode="json", exclude_none=True)
        ),
    }


@router.get("/alert-templates")
async def list_alert_templates() -> dict[str, Any]:
    """Every shipped template this build knows (SPEC §45, ``docs/API.md`` §5).

    Static data — the catalogue is code, not a table — so this takes no
    database at all. What a user has *enabled* is
    ``alerts.enabled_templates`` in the configuration, read through
    ``GET /api/internal/config``; what has been *instantiated* is the rule list
    below, where a shipped rule carries its ``template_key``.
    """
    return {"templates": [_template_payload(template) for template in SHIPPED_TEMPLATES]}


@router.get("/alert-rules")
async def list_alert_rules(request: Request) -> dict[str, Any]:
    """Every alert rule, by id."""
    rules = await _service(request).list_rules()
    return {"rules": [_rule_payload(rule) for rule in rules]}


@router.post("/alert-rules", status_code=status.HTTP_201_CREATED)
async def create_alert_rule(
    request: Request, body: Annotated[AlertRuleWriteRequest, Body()]
) -> dict[str, Any]:
    """Create a rule. The engine's rule set is recompiled before this returns."""
    try:
        record = await _service(request).create_rule(
            name=body.name,
            description=body.description,
            severity=body.severity,
            conditions=body.conditions,
            enabled=body.enabled,
        )
    except AlertRuleValueError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    return _rule_payload(record)


@router.put("/alert-rules/{rule_id}")
async def update_alert_rule(
    request: Request, rule_id: int, body: Annotated[AlertRuleWriteRequest, Body()]
) -> dict[str, Any]:
    """Replace a rule's definition. A full replace, not a patch.

    ``template_key`` is not replaceable: provenance is a statement about where
    a rule came from, and tuning a shipped rule does not make it stop having
    been shipped.
    """
    try:
        record = await _service(request).update_rule(
            rule_id,
            name=body.name,
            description=body.description,
            severity=body.severity,
            conditions=body.conditions,
            enabled=body.enabled,
        )
    except AlertRuleValueError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    except AlertRuleNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return _rule_payload(record)


@router.delete("/alert-rules/{rule_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_alert_rule(request: Request, rule_id: int) -> None:
    """Delete a rule and the matches it produced.

    The matches go with it because ``alert_matches.rule_id`` has no
    ``ON DELETE`` action (``docs/DATA_MODEL.md`` §4.3) and foreign keys are
    enforced (ADR-0001). Less is lost than that suggests: the sightings keep
    their ``max_alert_severity`` and their ``alert_matched`` events, and the
    activity feed keeps every ``alert_triggered`` row with the rule's name and
    the reason it gave.
    """
    if not await _service(request).delete_rule(rule_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=f"no alert rule with id {rule_id}")
